### ISSUE TITLE
we need syntax for choosing setting a property vs attribute on elements

### ISSUE THREAD

**Issue by trusktr** [preactjs/preact#4416]
- [ ] Check if updating to the latest Preact version resolves the issue

**Describe the bug**

It is not possible to guess 100% of the time whether an element user needs to set a property, or an attribute.

Does this set an attribute, or a property?

```jsx
return <some-element foo={[1, 2, 3]} />
```

Answer: it depends, sometimes Preact sets the `.foo` property, sometimes Preact sets the `foo` attribute. The user has no control over this.

The implementation of an element can vary (and may be out of the Preact user's control):

- some elements handle properties only
- some elements handle attributes only
- some elements handle both
- even if an element handles one or the other, or both
  - a user might want to set the attribute
  - or the property
  - or both
- user might want to add or remove an attribute based on a boolean (regardless if they also want to set the corresponding JS property or not)

It is not possible for Preact to always guess correctly which of the above cases the user will want, and the Preact user needs control.

**To Reproduce**

...

**Expected behavior**

Users need a way to express whether they want to set an attribute, or a property, or both. We also need to be able to add/remove attributes based on booleans.

Lit, Solid.js, and others have (sometimes only partial) solutions to this. For more details on the solutions other frameworks have, see this new test proposal for `custom-elements-everywhere`:

- https://github.com/webcomponents/custom-elements-everywhere/issues/2352

**Possible solution idea**

Solid.js has JSX syntax such as `attr:foo=...` and `prop:foo=...` for setting an attribute vs a property, but no boolean support. Pota, a library that is built on Solid.js, has its own JSX and `html` syntax that adds `bool:foo=...` support. Etc.

Preact could choose to adopt a similar approach, which is valid JSX syntax.

An alternative `html` template string tag for Preact could also provide Lit-like syntax (`.foo=...`, `foo=...`, and `?foo=...`).

**Comment by rschristian** [comment]
> Answer: it depends, sometimes Preact sets the `.foo` property, sometimes Preact sets the `foo` attribute. The user has no control over this.

While I don't disagree with the premise/request, and have written up minor forks/patches that add this, there are quite defined rules. It's not random or a guess: if the provided label has a DOM property, and isn't from a [small list of excluded labels](https://github.com/preactjs/preact/blob/4c20c23c16dd60f380ce9fe98afc93041a7e1562/src/diff/props.js#L107-L119), it gets set as a property, else, attr.

**Comment by trusktr** [comment]
> It's not random or a guess: if the provided label has a DOM property, and isn't from a [small list of excluded labels](https://github.com/preactjs/preact/blob/4c20c23c16dd60f380ce9fe98afc93041a7e1562/src/diff/props.js#L107-L119), it gets set as a property, else, attr.

That's true, "guess" could be considered in correct. The logic picks something based on conditions. Perhaps heuristic or rules are better terms.

To re-express with those words, no heuristic or rule can satisfy all scenarios because all scenarios can only be satisfied by the user picking which to use.

**Comment by rschristian** [comment]
Not quite what I meant, no. We largely set what we set, we're not looking to satisfy any scenarios beyond "works with DOM APIs correctly". Anything beyond that is a non-goal of the project really, something for users to address. It's not a concern of the project if it doesn't enable a CSS selector, for example.

Not having great tools for users to address these issues is, however, a valid issue for the project, if that makes sense.

Minor (perhaps frivolous) semantics, I know, but I want to clarify project goals.

**Comment by trusktr** [comment]
"works with DOM" and "allow the user to fully control DOM" are two different things. Note that being able to set choose between setting an element's JS property, or an element's DOM attribute, are two fundamentally required capabilities all developers working with DOM need. Based on that, I believe it is an excellent goal for a DOM framework like Preact to allow its users to have control over DOM in certain necessary ways.

So, based on the above, it seems you might be saying it is not Preact's goal to fix this because "works with DOM" is good enough. But that's not good enough for people working with DOM (which includes Custom Elements).

And yeah, enabling a CSS selector has absolutely nothing to do with Preact. But what Preact is doing is disallowing a user from setting an attribute within the Preact framework (which may happen to be used for a selector, or for something else like data binding with libraries external to Preact).

If I understand correctly, are you summarizing that users should escape out of Preact by using refs to go ahead and non-declaratively set DOM properties or attributes?

**Comment by rejhgadellaa** [comment]
I recently ran into the `popover` property not accepting boolean values, but the attribute can be a boolean (it will then default to `auto`).  So this:

```jsx
<menu popover />
```

Will actually try to set `Element.popover = true` (which is an error)

You have to write:

```jsx
<menu popover="auto" />
```

Not a big problem, but annoying 😅

**Comment by rschristian** [comment]
@rejhgadellaa have you checked the latest version? I believe that was a bug that was corrected.

**Comment by rschristian** [comment]
@trusktr I'm saying the comments of "It is not possible for Preact to always guess correctly which of the above cases the user will want" and "no heuristic or rule can satisfy all scenarios" aren't really relevant as it's not, nor was it ever, a project goal to do that. That was always on the user's end of things.

I just want to make that clear as the inability to "guess which cases the users will want" is not in itself a problem or a valid issue, as we never made any attempt at all to guess what a user could prefer in sone scenario. Lack of tools for the user to set things how they want, however, is absolutely a valid issue.

It's somewhat important framing, as it's confusing fundamental project goals, but I digress.

**Comment by hesxenon** [comment]
why not check the static `observedAttributes`? If my custom element is interested in e.g. `disabled` I must add "disabled" to the observedAttributes property anyway - which should sufficiently state that I am indeed interested in the content attribute and preact should not just drop that

**Comment by OEvgeny** [comment]
> why not check the static `observedAttributes`? If my custom element is interested in e.g. `disabled` I must add "disabled" to the observedAttributes property anyway - which should sufficiently state that I am indeed interested in the content attribute and preact should not just drop that

There may be attributes one does not want to observe. Better to have a way from Preact/JSX to denote either the attribute or property needs to be set as this sometimes can influence the behavior of particular custom element.

In addition developer using Preact may not have control over the Custom Element code, so they won't be able to adjust. Even extending a custom element class may not be an option in certain cases.

**Comment by hesxenon** [comment]
Okay, so basically I already worked around my issue by reflecting the IDL attributes to the content attribute, as hinted towards in the MDN docs (.value seems to be the exception, not the norm) and I don't have anything against the sigils approach, but your answer is a bit paradoxical...

> why not check the static observedAttributes?

> There may be attributes one does not want to observe

If a CE Author does not want to observe a content attribute they should not put it in something that's _literally called `observedAttributes`_.

If a CE defines that it _is_ interested in changes to the content attributes by listing it there it stands to reason that it will also react in some way to changes in those attributes (presumably by updating the IDL attribute, as hinted to in the MDN docs).

A developer that does not have control over the CEs code has nothing to do with it, no preact dev has control over the native elements code either yet we can easily use them in plain HTML (content) and preact (IDL).

Again, if preact/react would rather go the sigils route that's fine by me but saying that the semantics of `observedAttributes` change just because preact/react is being used seems... wrong.

**Comment by OEvgeny** [comment]
I mean Preact shouldn't rely solely on the particular CE `observedAttributes` static field to determine how to set a given property/attribute correctly. There always should be a way to adjust the behavior from JSX.

> If a CE Author does not want to observe a content attribute they should not put it in something that's literally called observedAttributes.

That's what I'm talking about. If the attribute isn't present in the array, this doesn't mean it should be treated as a property.

Moreover unobserved attributes can and may change in particular cases and there might be reasons for CE developer to refrain from the API usage.
